includes/class-wc-gateway-swedbank-pay-cc.php

```diff
@@ -570,7 +568,140 @@ public function swedbank_card_store() {
 	 * @return void
 	 */
 	public function thankyou_page( $order_id ) {
-		//
+		$order = wc_get_order( $order_id );
+		if ( ! $order ) {
+			return;
+		}
+
+		if ( $order->get_payment_method() !== $this->id ) {
+			return;
+		}
+
+		$payment_id = $order->get_meta( '_payex_payment_id' );
+		if ( empty( $payment_id ) ) {
+			return;
+		}
+
+		// Check tokens that should be saved or replaced
+		if ( '1' === $order->get_meta( '_payex_replace_token' ) ) {
+			try {
+				$result = $this->core->fetchPaymentInfo( $payment_id, 'authorizations,verifications' );
+
+				// Check payment state
+				switch ( $result['payment']['state'] ) {
+					case 'Ready':
+						// Replace token for:
+						// Change Payment Method
+						// Orders with Zero Amount
+						// Prepare sources where can be tokens
+						$sources = array();
+						if ( isset( $result['payment']['verifications'] ) ) {
+							$sources = array_merge( $sources, $result['payment']['verifications']['verificationList'] );
+						}
+
+						if ( isset( $result['payment']['authorizations'] ) ) {
+							$sources = array_merge($sources, $result['payment']['authorizations']['authorizationList']);
+						}
+
+						foreach ( $sources as $source ) {
+							$payment_token    = $source['paymentToken'];
+							$recurrence_token = $source['recurrenceToken'];
+							$card_brand       = $source['cardBrand'];
+							$masked_pan       = $source['maskedPan'];
+							$expiry_date      = explode( '/', $source['expiryDate'] );
+
+							// Create Payment Token
+							$token = new WC_Payment_Token_Swedbank_Pay();
+							$token->set_gateway_id( $this->id );
+							$token->set_token( $payment_token );
+							$token->set_recurrence_token( $recurrence_token );
+							$token->set_last4( substr( $masked_pan, - 4 ) );
+							$token->set_expiry_year( $expiry_date[1] );
+							$token->set_expiry_month( $expiry_date[0] );
+							$token->set_card_type( strtolower( $card_brand ) );
+							$token->set_user_id( get_current_user_id() );
+							$token->set_masked_pan( $masked_pan );
+
+							// Save Credit Card
+							$token->save();
+
+							// Replace token
+							delete_post_meta( $order->get_id(), '_payex_replace_token' );
+							delete_post_meta( $order->get_id(), '_payment_tokens' );
+							$order->add_payment_token( $token );
+
+							wc_add_notice( __( 'Payment method was updated.', 'swedbank-pay-woocommerce-payments' ) );
+
+							break;
+						}
+					default:
+						// no default
+				}
+			} catch ( Exception $e ) {
+				$this->adapter->log(
+					LogLevel::WARNING, sprintf( '%s::%s %s', __CLASS__, __METHOD__, $e->getMessage() )
+				);
+			}
+		}
+
+		// Wait for order status update by a callback
+		$status = $order->get_status();
```

Comment:
@asbjornu Because that's reason of the problem. When we get failed transaction then order status has failed state (Order has been updated by the callback). When customer continue payment with correct details and make succesfull payment then he or she will redirects to "order recieved" page. In this moment woocommerce checks an order status and if it's failed then shows "payment was unsuccessful." as was described [there](https://github.com/SwedbankPay/swedbank-pay-woocommerce-payments/issues/20). See [woocoommece sources](https://github.com/woocommerce/woocommerce/blob/master/templates/checkout/thankyou.php#L29).

Reply:
@asbjornu In other words, we need to know that the callback changed the order status. If it doesnt' work then do it itself.

Reply:
Just to make me understand: Is the problem that WooCommerce orders are reused for several payments and the status of a previously failed payment will mean the WooCommerce order status itself is failed? And based on the order status, WooCommerce will redirect to different pages?

Reply:
> Just to make me understand: Is the problem that WooCommerce orders are reused for several payments and the status of a previously failed payment will mean the WooCommerce order status itself is failed? And based on the order status, WooCommerce will redirect to different pages?

You're right

Reply:
Ok. But won't this introduce a 1 minute load time for all customers being redirected to the `thankyou_page`?

Reply:
@asbjornu There's just workaround for merchants which have problems with callbacks from swedbank-pay. I i think there won't be any problems.